client/core/status.go

```diff
@@ -0,0 +1,382 @@
+// This code is available on the terms of the project LICENSE.md file,
+// also available online at https://blueoakcouncil.org/license/1.0.0.
+
+package core
+
+import (
+	"fmt"
+
+	"decred.org/dcrdex/client/asset"
+	"decred.org/dcrdex/dex/msgjson"
+	"decred.org/dcrdex/dex/order"
+)
+
+// statusResolutionID is just a string with the basic information about a match.
+// This is used often while logging during status resolution.
+func statusResolutionID(dc *dexConnection, trade *trackedTrade, match *matchTracker) string {
+	return fmt.Sprintf("host = %s, order = %s, match = %s", dc.acct.host, trade.ID(), match.id)
+}
+
+// resolveMatchConflicts accepts a dexConnection and a slice of match status
+// conflicts. resolveMatchConflicts will block until the call to match_status
+// returns, but then goroutines are started to handle the actual resolution.
+// The trackedTrades' mutexes will remain locked until resolution completes.
+func (c *Core) resolveMatchConflicts(dc *dexConnection, statusConflicts []*matchStatusConflict) {
+
+	statusRequests := make([]*msgjson.MatchRequest, 0, len(statusConflicts))
+	// Lock the trade mutexes now until we're done. The other option would be
+	// to lock the trade mutexes in resolveConflictWithServerData, but that
+	// might allow a tick to occur between the match_status request and
+	// resolveConflictWithServerData. Our resolvers never send any requests or
+	// or responses. The longest running functions only request some wallet
+	// data, so I wouldn't expect these locks to be held for long. Regardless,
+	// we'll run each resolveConflictWithServerData below as a goroutine just to
+	// make sure we're not stacking those wallet calls.
+	for _, conflict := range statusConflicts {
+		conflict.trade.mtx.Lock()
```

Comment:
I figured this would raise some flags. See [my other reply](https://github.com/decred/dcrdex/pull/704#discussion_r498343689) on the reasoning for this synchronization. I'm not crazy about the pattern either, but I don't see anywhere it would fail. The complicating factor is that we want to retrieve all of the match data for all matches in a single `match_status` request, but we want the mutexes to be locked before we do that. 

I'm certainly open to other patterns as long as we're not changing that behavior.

Reply:
> as we're not changing that behavior.

Or as long as we can handle the race, I guess. Maybe we could lock in `resolveConflictWithServerData` instead, and then compare the server status from `connect` response to the server status from the `match_status` response, and skip resolution if it has changed. Not sure what to do when we use `resolveMatchConflicts` from anywhere other than `authDEX` though.